File: Knapsack/knapsack.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <stack>
#include <iomanip>

#define inputFile "knapsack.inp"
#define outputFile "knapsack.out"

using namespace std;

int n; // number of items
int weightLimit;

vector<int> weight;
vector<int> value;

// d[i][w] stores the max total value of items from 1 to n, in which w is the limit of weight
vector<vector<int>> d;

void Input()
{
    ifstream f;
    f.open(inputFile);

    f >> n >> weightLimit;

    weight.resize(n + 1, 0);
    value.resize(n + 1, 0);

    for (int i = 1; i < n + 1; ++i)
    {
        f >> weight[i] >> value[i];
    }

    f.close();
}



void Process()
{
    // init
    // n + 1 rows: denotes items
    // weightLimit + 1 columns: denotes limit of weight
    d.resize(n + 1, vector<int>(weightLimit + 1, 0));

    for (int i = 1; i < n + 1; ++i)
    {          
        for (int w = 1; w < weightLimit + 1; ++w)
        {
            // suppose item i-th is not selected
            d[i][w] = d[i - 1][w];

            // re-assign d[i][w] when item i-th will be selected for better total value
            if (weight[i] <=  w)
            {
                d[i][w] = max(d[i][w], d[i - 1][w - weight[i]] + value[i]);
            }
        }
    }
}

void Output()
{
    stack<int> itemStack; // contains items in order
    
    int wl = weightLimit; // temp weight limit
    int i = n; // tmp number of an item
    while (i)
    {
        // The weight changes due to item i
        if (d[i][wl] != d[i - 1][wl])
        {
            itemStack.push(i);
            wl = wl - weight[i];
        }
        --i;
    }

    ofstream f;
    f.open(outputFile);

    f << d[n][weightLimit] << '\n';

    while (!itemStack.empty())
    {
        int item = itemStack.top();
        f << item << ' ' << weight[item] << ' ' << value[item];
        if (itemStack.size() > 1) f << '\n';
        itemStack.pop();
    }

    f.close();
}

// show table
void show()
{
    // display column titles
    cout << string(6 + 2, ' ');
    for (int c = 0; c < weightLimit + 1; ++c)
    {
        cout << setw(6) << c;
    }
    cout << endl;

    // display a seperate line
    cout << string(6 + 2, ' ');    
    cout << string((weightLimit + 1) * 6, '-') << endl;
 
    // display row titles and values
    for (int r = 0; r < n + 1; ++r)
    {
        cout << setw(6) << r << " |";
        for (int c = 0; c < weightLimit + 1; ++c)
        {
            cout << setw(6) << d[r][c];
        }
        cout << endl;
    }
}

int main()
{
    Input();
    Process();
    show();      
    Output();
    return 0;
}

File: Knapsack/knapsack.py
import os
import sys
from prettytable import PrettyTable

input_file = os.path.join(sys.path[0], 'knapsack.inp')
output_file = os.path.join(sys.path[0], 'knapsack.out')

#-----------------------------------------------------------------------------
def input_data():
    global n, weight_limit
    global weight, value
    weight = list()
    value = list()

    with open(input_file) as f:
        n, weight_limit = map(int, f.readline().split())

        weight.append(0) # the zero-th element is not used
        value.append(0) # the zero-th element is not used
        for line in f.readlines():
            w, v = map(int, line.split())
            weight.append(w)
            value.append(v)

#-----------------------------------------------------------------------------
def process():
    # d[i][w] stores the max total value of items from 1 to n,
    # in which w is the limit of weight
    global d
    d = list()

    # init
    # n + 1 rows: denotes items
    # weight_limit + 1 columns: denotes limit of weight
    d = [[0 for col in range(weight_limit + 1)] for row in range(n + 1)]

    for i in range(1, n + 1):
        for w in range(1, weight_limit + 1):
            # suppose item i-th is not taken
            d[i][w] = d[i - 1][w]

            # re-assign d[i][w] when item i-th will be selected for better total value
            if weight[i] <= w:
                d[i][w] = max(d[i][w], d[i - 1][w - weight[i]] + value[i])

#-----------------------------------------------------------------------------
def output():
    s = list() # this is used as a stack

    wl = weight_limit # temp weight limit
    i = n # temp number of an item
    while i:
        if not d[i][wl] == d[i - 1][wl]:
            s.append(i)
            wl = wl - weight[i]
        i -= 1

    with open(output_file, 'w') as f:
        f.write(f'{d[n][weight_limit]}\n')
        for idx, item in enumerate(reversed(s)):
            line = f'{item} {weight[item]} {value[item]}'
            f.write(line) if idx == len(s) - 1 else f.write(line + '\n')

#-----------------------------------------------------------------------------
def show_table():
    pt = PrettyTable()

    # column header
    column_header = [w for w in range(weight_limit + 1)]
    column_header.insert(0, '') # insert blank cell
    pt.field_names = column_header

    # row
    for i in range(n + 1):
        row = [x for x in d[i]]
        row.insert(0, i) # row header
        pt.add_row(row)
    print(pt)

#-----------------------------------------------------------------------------
if __name__ == '__main__':
    input_data()
    process()
    show_table()
    output()